addons/payment_payumoney/models/payumoney.py

```diff
@@ -0,0 +1,185 @@
+
+_logger = logging.getLogger(__name__)
+
+
+class AcquirerPayuMoney(models.Model):
+    _inherit = 'payment.acquirer'
+
+    def _get_payu_urls(self, environment):
+        """ PayuMoney URLs
+        """
+        if environment == 'prod':
+            return {'payu_form_url': 'https://secure.payu.in/_payment'}
+        else:
+            return {'payu_form_url': 'https://test.payu.in/_payment'}
+
+    @api.model
+    def _get_providers(self):
+        providers = super(AcquirerPayuMoney, self)._get_providers()
+        providers.append(['payumoney', 'PayuMoney'])
+        return providers
+
+    merchant_id = fields.Char(string='Merchant Key', required_if_provider='payumoney')
+    payu_salt = fields.Char(string='Salt', required_if_provider='payumoney')
+
+    def _payu_generate_sign(self, acquirer, inout, values):
+        """ Generate the shasign for incoming or outgoing communications.
+        :param browse acquirer: the payment.acquirer browse record. It should
+                                have a shakey in shakey out
+        :param string inout: 'in' (odoo contacting payumoney) or 'out' (payumoney
+                             contacting odoo).
+        :param dict values: transaction values
+
+        :return string: shasign
+        """
+        assert inout in ('in', 'out')
+        assert acquirer.provider == 'payumoney'
+        def get_value(key):
+            if values.get(key):
+                return values[key]
+            return ''
+
+        if inout == 'in':
+            keys = "key|txnid|amount|productinfo|firstname|email|udf1|||||||||".split('|')
+            sign = ''.join('%s|' % (get_value(k)) for k in keys)
+            sign += acquirer.payu_salt
+        else:
+            keys = "|status||||||||||udf1|email|firstname|productinfo|amount|txnid".split('|')
+            sign = ''.join('%s|' % (get_value(k)) for k in keys)
+            sign = acquirer.payu_salt + sign + acquirer.merchant_id
+
+        shasign = hashlib.sha512(sign).hexdigest()
+        return shasign
+
+    @api.multi
+    def payumoney_form_generate_values(self, partner_values, tx_values):
+        self.ensure_one()
+        base_url = self.env['ir.config_parameter'].get_param('web.base.url')
+        payu_tx_values = dict(tx_values)
+        payu_tx_values.update({
+            'key': self.merchant_id,
+            'txnid': tx_values['reference'],
+            'amount': tx_values['amount'],
+            'productinfo': tx_values['reference'],
+            'firstname': partner_values['first_name'],
+            'email': partner_values['email'],
+            'phone': partner_values['phone'],
+            'service_provider': 'payu_paisa',
+            'surl': '%s' % urlparse.urljoin(base_url, PayuMoneyController._return_url),
+            'furl': '%s' % urlparse.urljoin(base_url, PayuMoneyController._exception_url),
+            'curl': '%s' % urlparse.urljoin(base_url, PayuMoneyController._cancel_url),
+        })
```

Comment:
Don't do a separate `udpate` call (least of all by passing it a literal dict), just set all these as keyword arguments on the original `dict` call:
```python
payu_tx_value = dict(tx_values,
    key=self.merchant_id,
    txnid=tx_values['reference'],
    …
```